snuba/query/expressions.py

```diff
@@ -0,0 +1,217 @@
+        sequences, this acts on a subtree. The semantics of transform can be different
+        between intermediate nodes and leaves, so each node class can implement it
+        its own way.
+
+        It returns the new expression. It does not guarantee the returned value
+        will be a new object. No guarantee of immutability is given on expressions.
+        """
+        raise NotImplementedError
+
+
+class HierarchicalExpression(Expression, NodeContainer[Expression]):
+    """
+    Expression that represent an intermediate node in the tree, which thus
+    can have children.
+
+    It provides two methods to iterate and transform. They both traverse
+    the subtree in a postfix order.
+    """
+
+    def transform(self, func: Callable[[Expression], Expression]) -> Expression:
+        """
+        Transforms the subtree starting from the children and then applying
+        the transformation function to the root.
+        This order is chosen to make the semantics of transform more meaningful,
+        the transform operation will be performed on thechildren first (think
+        about the parameters of a function call) and then to the node itself.
+
+        The consequence of this is that, if the transformation function replaces
+        the root with something else, with different children, we trust the
+        transformation function and we do not run that same function over the
+        new children.
+        """
+        self._set_children(
+            list(map(lambda child: child.transform(func), self._get_children()))
+        )
+        return func(self)
+
+    def __iter__(self) -> Iterator[Expression]:
+        """
+        Traverse the subtree in a postfix order.
+        The order here is arbitrary, postfix is chosen to follow the same
+        order we have in the transform method.
+        """
+        for child in self._get_children():
+            if isinstance(child, HierarchicalExpression):
+                for sub in child:
+                    yield sub
+            else:
+                yield child
+        yield self
+
+    @abstractmethod
+    def _get_children(self) -> Sequence[Expression]:
+        """
+        To be implemented by the subclasses to provide the list of children
+        for iteration/transformation.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def _set_children(self, children: Sequence[Expression]) -> None:
+        """
+        To be implemented by the subclasses to reset the list of children
+        for iteration/transformation.
+        """
+        raise NotImplementedError
+
+
+@dataclass
+class Null(Expression):
+    """
+    SQL NULL
+    """
+
+    def format(self) -> str:
+        raise NotImplementedError
+
+    def transform(self, func: Callable[[Expression], Expression]) -> Expression:
+        return self
+
+
+@dataclass
+class Literal(Expression):
+    """
+    A literal in the SQL expression
+    """
+    value: str
```

Comment:
I suspect that this will end up being a union type that includes other types (e.g. numerics.)

Is there any reason to make a distinction between `Literal(None)` and `Null()`? That seems redundant at this point.

Reply:
No there is not. Will merge

Reply:
I don't think we need the sequence of literals, because, like other conditions it will be expressed as a function:

```
5c4070ddfe70 :) SELECT event_id from sentry_local where in(project_id, tuple(2, 3));

SELECT event_id
FROM sentry_local
WHERE project_id IN (2, 3)

Ok.
```

Moreover in() accepts expressions. Not that it would be a good idea to do that.
I think we should add meaningful subclasses of FunctionCall when we start using it to reduce boilerplate.

Reply:
Good point — I hadn't considered this, I guess as long as we have all the scalar types, we should be able to use `array` and/or `tuple` to do everything else we need.

Looking at the [ClickHouse data types documentation](https://clickhouse.yandex/docs/en/data_types/) and considering our usage, we might want to consider adding UUID and Date/DateTime types? I'm not sure how these would be handled at the parser level to disambiguate them from plain strings, though…

Reply:
Unfortunately, the UUID comes as a string from the client, the parser does not know whether it should be translated into a clickhouse UUID or to be kept as a string. Datetime can be managed.

Reply:
I wonder if it'd make sense (at some point — not necessarily now) to make `value` a type variable so that the Snuba query and the ClickHouse (or whatever other table engine, per #582) query have different possible sets of literals. I can't think of any other ways that I'd expect the two different query structures to diverge significantly, at least at this point.

Thinking about it more, I'm not sure that introducing a DateTime type is a good thing here. I'd assume we'd have to do so using regular expressions?

Reply:
What about we wait when we start building some actual query processing to see what level of specialization makes sense for literals? I am not sure we should even have DateTime. There is no way in the query language to tell when a literal is actually a DateTime or a string containing a Datetime.

Reply:
I think that is fine.

> I am not sure we should even have DateTime. There is no way in the query language to tell when a literal is actually a DateTime or a string containing a Datetime.

For what it's worth, I think we're on the same page here.